### DISCUSSION TITLE
Acerca del trabajo parcial (post largo)

### DISCUSSION THREAD

**Discussion by mmasias** [mmasias/23-24-eda2#158] [General]
Luego de revisarlo, es posible hacer un fork de un fork y luego redirigir la contribución al repositorio original.

Entonces:

- El **usuarioA** puede hacer un fork del repositorio **mmasias/23-24-eda2** y luego de eso tiene **usuarioA/23-24-eda2**
- El **usuarioB** hace un fork del repositorio de **usuarioA**, con lo cual tendrá **usuarioB/23-24-eda2**
- Al momento de hacer sus contribuciones (pull request), GitHub permite elegir el repositorio de destino hacia el cual va dicho PR. 

De hecho, si ahora intentan hacer un PR al repo de la asignatura, verán como pueden elegir cualquiera de los repos de los compañeros

|![image](https://github.com/mmasias/23-24-eda2/assets/8528047/7b2a6e9f-af98-4ba0-9bda-d48ea5b6b4f3)|
|-|
|*En el ejemplo, desde la cuenta de ibuprofeno estoy haciendo un PR. Nótese como además de la rama de destino (que era lo que hemos hecho hasta hoy) puedo elegir cualquiera de los repos que han hecho fork del repo original*|

No obstante, el trabajo lo haremos y entregaremos en [este otro repositorio](https://github.com/mmasias/23-24-eda2-ep), por varias razones:

1. Por no contaminar el repositorio original de la asignatura.
2. Por tener una zona limpia de revisión, solo con el código a evaluar.

Investigando (ha venido bien la pregunta de clase), descubrí que **[existe una manera de mover el contenido de un repositorio a otro](https://blog.gbellmann.technology/2017/08/06/moviendo-algunos-archivos-de-un-repositorio-git-a-otro-conservando-la-historia/)**, manteniendo el historial de construcción (que es lo que queremos).

Por lo tanto, (y siguiendo las indicaciones de arriba):

1. Preparar el repo donde tienen su código, de modo que se mantenga el código original y su histórico.
2. Hacerlo apuntar al repo oficial del trabajo
3. Trabajar sobre ese repo, colaborando (en caso que decidan trabajar en grupo) con PR entre los miembros del grupo
4. Hacer el PR final al repo oficinal del trabajo.

**Comment by mmasias** [comment]
Confirmar la lectura del post anterior mediante ⬆️ y si hay dudas, plantearlas en este hilo.

**Comment by andriuw08** [comment]
Nosotros que somos dos hicimos esto, yo como el responsable forkeo el repo del parcial, clone mi repo de eda2, en visual code con mi repo ya abierto borre todas las carpetas menos entregas/andriuwYepez/reto-003, incluyendo los readme y todo, luego con la terminal de visualcode estando en mi repo meti los comandos que dio masias para preparar el repo y para redireccionar el repo, luego con el git remote -v veo que ya este apuntando al repo del parcial, cabe recalcar que en el comando que direcciona al repo en la parte de la url debemos meter la url que da el nuestro repo del parcial en clone, es como si lo fuesemos a clonar pero en lugar de clonar lo metemos ahi en la url, luego cuando ya este apuntando al repo del parcial, iniciamos sesion en git con los comando que se meten por primera vez para colocar usuario y correo electronico, luego creamos una rama diferente a la del main, hacemos commit, publicar rama y ya en nuestro repo deberia aparecer que hay cambios, junto a la rama y el historial de cambios. 

Ya luego el otro compañero hace lo mismo pero va a apuntar al repo principal cuando redirecciona a un url.
![Captura de pantalla 2024-03-21 113046](https://github.com/mmasias/23-24-eda2/assets/114088104/17cb9a67-3981-4218-9f0d-d2448ad640e0)

**Comment by mmasias** [comment]
Un apunte, con respecto a vocabulario:

- Repo: El repo de la asignatura en la cuenta de Manuel
- Fork: Un fork en la cuenta de algún alumno
- Clon: Una copia de un repo (sea fork o sea el repo) en un ordenador

Entonces, no es lo mismo decir "hago xxx en el clon" que "hago xxx al repo" que "hago xxx al fork". Es importante mantener esta convención al momento de describir las soluciones.

**Comment by AndrextA** [comment]
# Procedimiento para la Transferencia de una Rama de un Repositorio Existente a un Nuevo Repositorio

Para efectuar la transferencia de una rama existente de nuestro repositorio (Fork) principal a un nuevo repositorio (Repo), es necesario seguir una serie de pasos detallados a continuación.

## Primer Paso:
Es imprescindible tener una copia local (Clon) del repositorio en cuestión en nuestro ordenador. Esto nos permitirá ejecutar los comandos necesarios de Git.

![image](https://github.com/mmasias/23-24-eda2/assets/99576569/6e4ec002-1668-4708-98b7-04c76f058689)


## Segundo Paso
Es necesario añadir a nuestro clon (Fork) el nuevo repositorio (Repo) al que deseamos transferir los datos de nuestra rama existente. Para hacer esto, utilizamos el comando

```
git remote add nombreRepo URL
```

Por ejemplo:

```
git remote add parcial git@github.com:AndrextA/23-24-eda2-ep.git
```

Después de ejecutar este comando, para verificar que el repositorio se ha añadido correctamente, podemos usar el comando

```
git remote -v
```

Este comando nos mostrará todos los repositorios remotos que contiene nuestro repositorio local.

![image](https://github.com/mmasias/23-24-eda2/assets/99576569/02b4ba5d-63ac-42b9-9924-797ae1531143)


## Tercer Paso
 Finalmente, debemos realizar un ‘push’ al nuevo repositorio (Repo), especificando la rama y el nombre de la rama que se generará en el repositorio destino. Para hacer esto, utilizamos el comando
 ```
 git push nombreRepo nombreRamaExistente:nombreRama
 ```

Por ejemplo:

```
git push parcial evaluacionContinua3:parcial-eda
```

Una vez completados estos pasos, deberíamos ver la nueva rama en nuestro nuevo repositorio (Repo) con los datos que ya existían en nuestro repositorio anterior (Fork).

![image](https://github.com/mmasias/23-24-eda2/assets/99576569/70b3299c-b674-4f82-9690-da102f0a2420)
### Repositorio antes y después
![image](https://github.com/mmasias/23-24-eda2/assets/99576569/7f172d4f-d86b-4fad-ab54-1bc293f98860)
![image](https://github.com/mmasias/23-24-eda2/assets/99576569/d610f685-bd7b-4735-a014-f0a99a9b87a2)

**Comment by AndrextA** [comment]
Es importante mencionar un detalle adicional. Si intentan transferir directamente la rama del repositorio (Fork) a la rama ‘main’ del nuevo repositorio (Repo), encontrarán que esto no es posible.

![WhatsApp Image 2024-03-21 at 19 28 57_306a5ee0](https://github.com/mmasias/23-24-eda2/assets/99576569/4f02b0bb-5569-494a-b855-79666c22ec36)

 En su lugar, deben crear una nueva rama en el nuevo repositorio, que en los pasos anteriormente mecionados ya se realizo.
Posteriormente, pueden cambiar la rama por defecto en la configuración del repositorio. 

![chrome_as0LBjsDk4](https://github.com/mmasias/23-24-eda2/assets/99576569/b7e03f7a-ffb2-4d2a-a3c5-30966195efee)
![chrome_oEJNQys4Cw](https://github.com/mmasias/23-24-eda2/assets/99576569/facf1c36-dc3c-4be0-8fc5-606c8a0dbc05)
![image](https://github.com/mmasias/23-24-eda2/assets/99576569/dab30dff-51a0-4478-98c6-a20920ac4f7c)

 Si lo desean, también pueden cambiar el nombre de la rama y eliminar la rama  ‘main’  antigua.

**Reply by AndrextA** [reply]
Si les da un error por historias no relacionadas en git, usen estos comandos


git pull origin main --allow-unrelated-histories
git push -u origin main

Aquí el video explicativo:
https://www.youtube.com/watch?v=_leKT1LBhvc&ab_channel=TheLearningBird

**Comment by pablarce** [comment]
## Por si a alguien le sirve todavía, esta es la solución mas fácil y rápida que se me ha ocurrido en el caso de tener una rama `main` y rama `reto-003`:

### Primer paso:

- Hacer un fork del repo `23-24-eda2-ep`
- Clonar el repo `23-24-eda2`

### Segundo paso:

En el cmd del repo que acabas de clonar `23-24-eda2`

- `git fetch` (para disponer de las otras ramas antes de borrar la referencia)
- `git remote remove origin` en la ruta de la carpeta de `23-24-eda2` (importante para no romper este repo).
- Tambien puedes cambiar el nombre de la carpeta a `23-24-eda2-ep`, eso ya como quieras.

### Tercer paso:

#### SI tienes el contenido en la rama `main`:

- Simplemente borra todo* menos la carpeta de entregas, en la rama `main`. Despues haz un `git add .` y un `git commit -m""` para guardar estos cambios. 

#### SI tienes el contenido en la rama `reto-003`:

- Borra todo* el contenido de la rama `main`. Despues haz un `git add .` y un `git commit -m""` para guardar estos cambios.  
- haz `git checkout` a la rama `reto-003`, y repite el paso anterior, pero esta vez deja la carpeta entregas.
- En la rama `main`, haz `git merge reto-003`. Si lo has hecho bien no te debería dar problema. 

*no borrar el archivo .git.

### Cuarto paso:

En la rama `main` ahora actualizada:

1. `git remote add origin https://github.com/usuario/23-24-eda2-ep.git`
2. `git fetch`
3. `git pull origin main --allow-unrelated-histories`
4. `git push origin main`

Si haces un `git remote -v` vas a ver que lo tienes enlazado correctamente al fork de  `https://github.com/usuario/23-24-eda2-ep.git`

#### Si sigues estos paso te va a permitir conservar tantos tus commits de tu repo `23-24-eda2`, y tambien los commits del  `23-24-eda2-ep` original.
